dsbapi/__init__.py

```diff
@@ -7,20 +7,53 @@
 import uuid
 import base64
 
+DATA_URL = "https://app.dsbcontrol.de/JsonHandler.ashx/GetData"
+
 class DSBApi:
     def __init__(self, username, password):
         self.username = username
         self.password = password
 
+    # Sends a data request to the server.
+    # Returns the URL to the timetable HTML page
     def fetch_api(self):
+        # Iso format is for example 2019-10-29T19:20:31.875466
         current_time = datetime.datetime.now().isoformat()
-        current_time = current_time.split(".")[0] + "." + current_time.split(".")[1][:3] + "Z"
-        innerDATA = {"UserId":self.username,"UserPw":self.password,"AppVersion":"2.5.9","Language":"de","OsVersion":"28 9","AppId":str(uuid.uuid4()),"Device":"SM-G935F","BundleId":"de.heinekingmedia.dsbmobile","Date":current_time,"LastUpdate":current_time}
-        jsonStream = io.BytesIO()
-        with gzip.open(filename=jsonStream, mode='wt') as streamReader:
-            streamReader.write(json.dumps(innerDATA, separators=(',', ':')))
-        timetable_data = requests.request("POST", "https://app.dsbcontrol.de/JsonHandler.ashx/GetData", json = {"req": {"Data": base64.encodebytes(jsonStream.getvalue()).decode("utf-8"), "DataType": 1}})
-        return [o["Root"]["Childs"][0]["Childs"][0]["Detail"] for o in json.loads(gzip.decompress(base64.b64decode(json.loads(timetable_data.text)["d"])))["ResultMenuItems"][0]["Childs"] if o["MethodName"] == "timetable"][0]
+        # Cut off last 3 digits and add 'Z' to get correct format
+        current_time = current_time[:-3] + "Z"
+        
+        # Parameters required for the server to accept our data request
+        params = {
+            "UserId": self.username,
+            "UserPw": self.password,
+            "AppVersion": "2.5.9",
+            "Language": "de",
+            "OsVersion": "28 9",
+            "AppId": str(uuid.uuid4()),
+            "Device": "SM-G935F",
+            "BundleId": "de.heinekingmedia.dsbmobile",
+            "Date": current_time,
+            "LastUpdate": current_time
+        }
+        # Convert params into the right format
+        params_bytestring = json.dumps(params).encode("UTF-8")
+        params_compressed = base64.b64encode(gzip.compress(params_bytestring)).decode("UTF-8")
```

Comment:
Ich bin unsicher ob diese JSON-Struktur (Seperators und das Encoding) funktionieren oder das ganze fehleranfälliger machen. Solche und noch kleinere Änderungen haben das gesamte Projekt gebrochen, weil die Betreiber davon wohl in ihren Logs Wind von bekommen haben. Ich würde lieber am Request bleiben, der dem der App am nächsten ist.

Reply:
Okay, dann ändere ich das ganze, so dass es vom Verhalten identisch zu vorher bleibt

Reply:
Wegen dem Encoding: aus
```python
jsonStream = io.BytesIO()
with gzip.open(filename=jsonStream, mode='wt') as streamReader:
    streamReader.write(json.dumps(innerDATA, separators=(',', ':')))
base64.encodebytes(jsonStream.getvalue()).decode("utf-8")
```
habe ich jetzt
```python
params_bytestring = json.dumps(params, separators=(',', ':')).encode("UTF-8")
base64.b64encode(gzip.compress(params_bytestring)).decode("UTF-8")
```
gemacht. Das zweite `decode("UTF-8")` ist also gleichgeblieben, darum müssen wir uns keine Sorgen machen.

Mein `encode("UTF-8")` habe ich eingefügt, weil `gzip.compress` mit Bytes statt Strings arbeitet. Ich glaube aber, der Original-Code enkodiert den Text auch als UTF-8, nur halt versteckt im Hintergrund. Im Original-Code wird nämlich ein gzip-Stream mit dem Modus 'wt' geöffnet wurde, also ein Textmodus. Der Textmodus ermöglicht, dass man Unicode in den Stream reinschreibt, und Bytes herausbekommt. Der Textmodus _muss_ also irgendwo Unicode in Bytes umwandeln.

Ich will damit sagen, dass auch der `encode("UTF-8")` Call schon in einer anderen Form im Original-Code vorhanden war, und man ihn deshalb auch nicht entfernen kann

(Hoffentlich war das verständlich)

Reply:
Ja, allerdings ist das Encoding nicht zwingend nötig,  da JSON auch Bytes nimmt. Aber gut.